polygeist/tools/cgeist/Lib/clang-mlir.cc

```diff
@@ -4789,8 +4810,28 @@ mlir::Value MLIRASTConsumer::GetOrCreateGlobalLLVMString(
   return globalPtr;
 }
 
-mlir::func::FuncOp MLIRASTConsumer::GetOrCreateMLIRFunction(
-    const FunctionDecl *FD, const bool ShouldEmit, bool getDeviceStub) {
+llvm::Optional<mlir::FunctionOpInterface>
+MLIRASTConsumer::getFunction(const std::string &name,
+                             FunctionContext context) const {
+  const auto find = [&](const auto &map) {
+    const auto Iter = map.find(name);
+    return Iter == map.end()
+               ? llvm::Optional<mlir::FunctionOpInterface>{}
+               : llvm::Optional<mlir::FunctionOpInterface>{Iter->second};
+  };
+  switch (context) {
+  case FunctionContext::Host:
+    return find(functions);
+  case FunctionContext::SYCLDevice:
+    return find(deviceFunctions);
+  }
+  llvm_unreachable("Invalid function context");
```

Comment:
return llvm::None here. The caller then will need to check whether the Optional has a value.

Reply:
In this case, I don't think we want to do that. If a new value is added to `FunctionContext`, I'd like this function to abort execution so that the programmer knows he should address the new case here. That code should in fact be unreachable.

Reply:
For the record: when a function return an `Optional<>` it means the function might return llvm::None. If an assert (or llvm_unreachable) is what you want then the function should not return an optional value.